components/family-studies/migrations/src/main/java/org/phenotips/studies/family/migrations/R71500PhenoTips2944DataMigration.java

```diff
@@ -0,0 +1,311 @@
+            return false;
+        }
+        return updatedGenes || updatedColors;
+    }
+
+    // convert from e.g.
+    // "colors":{"causalGenes":{"SLC37A4":"#eac080"},
+    // to
+    // "colors":{"causalGenes":{"ENSG00000137700":"#eac080"},
+    private boolean convertColors(JSONObject colorsJSON, Map<String, String> hgncToEnsemblMap)
+    {
+        boolean updated = false;
+        String[] geneFileds =
+        { PEDIGREE_SOLVED_GENES_FIELD, PEDIGREE_CANDIDATE_GENES_FIELD, PEDIGREE_REJECTED_GENES_FIELD };
+        for (String field : geneFileds) {
+            JSONObject colors = colorsJSON.optJSONObject(field);
+            if (colors == null || colors.length() == 0) {
+                continue;
+            }
+
+            for (String geneSymbol : colors.keySet()) {
+                String color = colors.optString(geneSymbol);
+                String enselmbId = hgncToEnsemblMap.containsKey(geneSymbol) ? hgncToEnsemblMap.get(geneSymbol)
+                    : this.getEnsemblId(geneSymbol);
+                colors.put(enselmbId, color);
+                colors.remove(geneSymbol);
+            }
+            updated = true;
+        }
+
+        return updated;
+    }
+
+    // convert from e.g.
+    // "genes":[{"gene":"SLC37A4","status":"candidate"}]
+    // to
+    // "genes":[{"gene":"ENSG00000137700","status":"candidate"}]
+    private boolean convertGenes(JSONObject nodeJSON, Map<String, String> hgncToEnsemblMap)
+    {
+        boolean updated = false;
+        JSONObject properties = nodeJSON.optJSONObject(PEDIGREE_PROPERTIES_STRING);
+        if (properties == null) {
+            return false;
+        }
+        JSONArray genes = properties.optJSONArray(PEDIGREE_GENES_FIELD);
+        if (genes == null || genes.length() == 0) {
+            return false;
+        }
+
+        for (Object gene : genes) {
+            JSONObject geneObj = (JSONObject) gene;
+            String geneSymbol = geneObj.optString(PEDIGREE_GENE_FIELD);
+            if (geneSymbol == null) {
+                continue;
+            }
+            String geneEnsemblId = this.getEnsemblId(geneSymbol);
+            hgncToEnsemblMap.put(geneSymbol, geneEnsemblId);
+            geneObj.put(PEDIGREE_GENE_FIELD, geneEnsemblId);
+            updated = true;
+        }
+
+        return updated;
+    }
+
+    /**
+     * Gets EnsemblID corresponding to the HGNC symbol.
+     *
+     * @param geneSymbol the string representation of a gene symbol (e.g. NOD2).
+     * @return the string representation of the corresponding Ensembl ID.
+     */
+    private String getEnsemblId(String geneSymbol)
+    {
+        final VocabularyTerm term = this.hgnc.getTerm(geneSymbol);
+        @SuppressWarnings("unchecked")
+        final List<String> ensemblIdList = term != null ? (List<String>) term.get("ensembl_gene_id") : null;
+        final String ensemblId = ensemblIdList != null && !ensemblIdList.isEmpty() ? ensemblIdList.get(0) : null;
```

Comment:
is `get(0)` always guaranteed to be the right term? E.g. is the first pone in array the best? What if there is No exact match, and the first is just the least accurate match? We may have free-text terms which have no exact match but do have some partial match

E.g. http://localhost:8080/get/PhenoTips/GeneNameService?outputSyntax=plain&rand=0.38996313003383465&json=true&q=GS

returns stuff, but none of it is the correct ensemble ID